nav2_bringup/bringup/launch/slam_launch.py

```diff
@@ -70,10 +73,26 @@ def generate_launch_description():
         description='Automatically startup the nav2 stack')
 
     # Nodes launching commands
+
+    # If the provided param file doesn't have slam_toolbox params, we must remove the 'params_file'
+    # LaunchConfiguration, or it will be passed automatically to slam_toolbox and will not load 
+    # the default file
+    has_slam_toolbox_params = HasNodeParams(source_file=params_file,
+                                            node_name='slam_toolbox')
+    # slam_toolbox with provided param_file
+    push_launch_config = PushLaunchConfigurations(
```

Comment:
If you have `has_slam_toolbox_params`, why not just use normal python `if else` to call `start_slam_toolbox_cmd` with and without the param file if it has it?

Reply:
Because I can't evaluate a LaunchConfiguration until it is executed.

If I understood correctly how `launch` works, there's two ways of processing a launch.py file, "describe" and "execute".
In the describe phase has_slam_toolbox_params does not evaluate to "True" or "False", but to a description of the substitution. Only on execution, it is able to resolve `params_file` into a file and see if it contains the node parameters.

Reply:
Maybe I'm missing something, but wouldn't 

```
    has_slam_toolbox_params = HasNodeParams(source_file=params_file,
                                            node_name='slam_toolbox')

    start_slam_toolbox_cmd = IncludeLaunchDescription(
        PythonLaunchDescriptionSource(slam_launch_file),
        condition=IfCondition(has_slam_toolbox_params),
        launch_arguments={'use_sim_time': use_sim_time, 'params_file', params_file}.items())
    start_slam_toolbox_cmd = IncludeLaunchDescription(
        PythonLaunchDescriptionSource(slam_launch_file),
        condition=UnlessCondition(has_slam_toolbox_params),
        launch_arguments={'use_sim_time': use_sim_time}.items())
```

work? `Only on execution, it is able to resolve params_file into a file and see if it contains the node parameters.` I'm not sure I see how that would make this not work? You would be executing this at runtime in all cases of python. The if/unless here would be analog to ROS1 XML as well and I think this should work, its how we do [`if/else` on slam vs localization](https://github.com/ros-planning/navigation2/blob/dcc1d4dafba6f6c7588cb5f7f6ca918b2a6c48ef/nav2_bringup/bringup/launch/bringup_launch.py#L92-L109)

Reply:
My first implementation was exactly like this.

But I faced the problem of the discussion above, that all this file LaunchConfigurations are passed automatically to included configurations, regardless of specifiying `'params_file', params_file`. So I had to do the UnsetLaunchConfiguration thing.

So the file was always being passed, regardless of the value of has_slam_toolbox_params.

Reply:
OK. With that in mind, can we move `start_map_saver_server_cmd` and `start_lifecycle_manager_cmd` before all this stuff so its at the end? 

I see you `UnsetLaunchConfiguration`, do you need to re-`LaunchConfiguration` it? I see how push/pop works. In your comment on L77, adding a sentence or two about what specifically is happening here would be really good for both my sanity in the future and anyone else reading this without context. It explains well why this is happening but doesn't talk about the push/pop.

Reply:
I've reordered and extended comments, please take a look when you have time.

And sorry about the mess, I also thought such a mess would not be required to pass a parameter optionally.